internal/kafka/internal/services/kafkatlscertmgmt/secure_storage.go

```diff
@@ -48,18 +51,37 @@ func newSecureStorage(config *config.AWSConfig, automaticCertificateManagementCo
 		return nil, err
 	}
 	return &secureStorage{
-		secretClient: secretClient,
-		secretCache:  secretCache,
-		secretPrefix: config.SecretManager.SecretPrefix,
+		secretClient:     secretClient,
+		secretCache:      secretCache,
+		storageLock:      sync.Mutex{},
+		storageItemLocks: map[string]*sync.Mutex{},
+		secretPrefix:     config.SecretManager.SecretPrefix,
 	}, nil
 }
 
 func (storage *secureStorage) Lock(ctx context.Context, key string) error {
-	return nil // NOOP as AWS Secret manager uses versioning mechanism to achieve optimistic locking
+	mu, ok := storage.storageItemLocks[key]
```

Comment:
Now that we manage the certs asynchronously, can happen that 2 threads call this method at the same time? If yes, here we could have a race condition.

Reply:
This will always be called in serial.

Reply:
Then what is the use of the lock?

Reply:
For a given `key`, the lock here helps guarding against processing of certificate management if the previous one for the given key hasn't finished. But for each given `key` there will be at least a `30s` interval between each cert management request.

Reply:
If It is not a thread synchronisation issue, it could be done with a simple check instead of a lock.

I would add a comment to the `storage` class describing that the class is not thread-safe (seeing the lock mechanisms could be misleading )

Reply:
Sorry, I misunderstood you comment about the `What is the use of the lock, then?`.
My answer on https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1639#discussion_r1137193186 was applicable for a key specific lock - which was highlighted by the first comment.
The other global lock `storage.storageLock.Lock()` (which I should probably rename) is prevent simultaneous access to the key specific locks. Re-reading everything again, you are right (and thank you for this) that there could be a race condition which might not have been covered by https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1639#discussion_r1137193186. 
This could typically happen for developer Kafkas that share the same certificate: this means that the key could be the same as well. Only standard kafkas have their separate certificate.
Now if we go to the developer Kafka case, it could happen that one developer kafka in `preparing` state and another in `ready` state have the certificate being reconciled at the same time: If this happens then we'll have a race condition.
The options: 
- Find unique certs and reconcile them in all our reconcilers
- Or make this thread safe

I'll go with the later option since we are at it and hopeful it should be quick (famous last word... :-)). The first option could be sort of covered by the second. And once we've the second option in, the storage could be used anywhere in a thread safe manner as long as `Lock` and `Unlock` are called.

Reply:
> The options:

> Find unique certs and reconcile them in all our reconcilers
> Or make this thread safe
> I'll go with the later option since we are at it and hopeful it should be quick (famous last word... :-)). The first option could be sort of covered by the second. And once we've the second option in, the storage could be used anywhere in a thread safe manner as long as Lock and Unlock are called.

Thinking about this, we might need the two options to be future proof. 
i.e we still need the storage to be thread safe.
and we'll need option one - which I'll refine as

Reply:
JIRA created https://issues.redhat.com/browse/MGDSTRM-10921

Reply:
@ziccardi this is done already. I'll resolve the comment. Thanks for raising this one.